client/app/components/QuerySelector.jsx

```diff
@@ -0,0 +1,160 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import { debounce, find } from 'lodash';
+import Input from 'antd/lib/input';
+import { Query } from '@/services/query';
+import { toastr } from '@/services/ng';
+import { QueryTagsControl } from '@/components/tags-control/TagsControl';
+
+const SEARCH_DEBOUNCE_DURATION = 200;
+
+class StaleSearchError extends Error {
+  constructor() {
+    super('stale search');
+  }
+}
+
+function search(term) {
+  // get recent
+  if (!term) {
+    return Query.recent().$promise
+      .then((results) => {
+        const filteredResults = results.filter(item => !item.is_draft); // filter out draft
+        return Promise.resolve(filteredResults);
+      });
+  }
+
+  // search by query
+  return Query.query({ q: term }).$promise
+    .then(({ results }) => Promise.resolve(results));
+}
+
+export function QuerySelector(props) {
+  const [searchTerm, setSearchTerm] = useState();
+  const [searching, setSearching] = useState();
+  const [searchResults, setSearchResults] = useState([]);
+  const [selectedQuery, setSelectedQuery] = useState();
+
+  let isStaleSearch = false;
+  const debouncedSearch = debounce(_search, SEARCH_DEBOUNCE_DURATION);
+  const placeholder = 'Search a query by name';
+  const clearIcon = <i className="fa fa-times" onClick={() => selectQuery(null)} />;
+  const spinIcon = <i className={cx('fa fa-spinner fa-pulse', { hidden: !searching })} />;
+
+  // set selected from prop
+  useEffect(() => {
+    if (props.selectedQuery) {
+      setSelectedQuery(props.selectedQuery);
+    }
+  }, [props.selectedQuery]);
+
+  // on search term changed, debounced
+  useEffect(() => {
+    // clear results, no search
+    if (searchTerm === null) {
+      setSearchResults(null);
+      return () => {};
+    }
+
+    // search
+    debouncedSearch(searchTerm);
+    return () => {
+      debouncedSearch.cancel();
+      isStaleSearch = true;
+    };
+  }, [searchTerm]);
+
+  function _search(term) {
+    setSearching(true);
+    search(term)
+      .then(rejectStale)
+      .then((results) => {
+        setSearchResults(results);
+        setSearching(false);
+      })
+      .catch((err) => {
+        if (!(err instanceof StaleSearchError)) {
+          setSearching(false);
+        }
+      });
+  }
+
+  function rejectStale(results) {
+    return isStaleSearch
+      ? Promise.reject(new StaleSearchError())
+      : Promise.resolve(results);
+  }
+
+  function selectQuery(queryId) {
+    let query = null;
+    if (queryId) {
+      query = find(searchResults, { id: queryId });
+      if (!query) { // shouldn't happen
+        toastr.error('Something went wrong... Couldn\'t select query');
+      }
+    }
+
+    setSearchTerm(query ? null : ''); // empty string triggers recent fetch
```

Comment:
It took some time for me to understand how it works, and then I realized that actually it's AngularJS-way to handle data flow: you set a state field, but setter just updates value; and there is watcher (`$scope.$watch`) that does something when that field changes. That's not something I personally expect to see in React code :thinking:

Reply:
I do it rarely when it makes sense. But now reconsidering more since useEffect is so convenient. Will stay away from it if you feel it's an anti-pattern.

Reply:
> Will stay away from it if you feel it's an anti-pattern.

Yes, I think so. That's what makes AngularJS code hard to understand; it's much easier to understand what's going on when you see: "ah, this function updates state variable, but also it loads queries", etc. With such side effects you see: "this function updates state. okay." - and then you see that oh magic! it somehow loads data! whaaat?? And you need to find another place in the code which load query results. And BTW - in AngularJS this mechanism is a bit better because you describe dependency first, and then write handler. With `useEffect` if's really easy to miss dependencies (what I did, actually). Compare:
```javascript
$scope.$watch('field', () => { 
  ...
  // a lot of code
  ...
});

useEffect(() => {
  ...
  // a lot of code
  ...
}, [field]); // <- this line left unattended
```

I really this is anti-pattern; probably one of the things I most liked in React - strict data flows, without implicit dependencies.

Reply:
Interesting observation.

Reply:
_(Read that code once more and have something to add)_

Another con of this approach is harder data flow control. Did you notice that effect that loads search results could be optimized? When search term is `null` - you already have search results (empty list). If you have setter for search term - you can handle it immediately:

1. user clicks "clear" button - search term set to `null`, and search results to `null`;
2. component renders because state changed.

 but with `useEffect` it will:

1. user clicks "clear" button - search term set to null;
2. component renders because state changed;
3. side effect executed, it sees that search term is `null` and updates search results to `null`
4. component renders because state changed.